diff --git a/omega/omega.theme b/omega/omega.theme
index f4f2ecb..c815747 100644
--- a/omega/omega.theme
+++ b/omega/omega.theme
@@ -248,26 +248,28 @@ function omega_preprocess_page(&$vars) {
   $layoutTheme = $layout['theme'];
   $layoutLayout = $layout['layout'];
   $layouts = omega_return_layouts($layoutTheme);
-  $region_groups = $layouts[$layoutLayout]['region_groups']['all'];
-  foreach($region_groups as $group_id => $group_info) {
-    $groupRegions = $region_groups[$group_id]['regions'];
-    // duplciate the #id into a class for better use with Omega.gs
-    $altered_group_id = str_replace("_", "-", $group_id);
-    $region_classes[$group_id][] = $altered_group_id . '-layout';
-    foreach($groupRegions as $region_id => $region_info) {
-      $altered_region_id = str_replace("_", "-", $region_id);
-      
-      if (isset($vars['page'][$region_id]['#region'])) {
-        $region_classes[$group_id][] = 'with--' . $altered_region_id;
-        $page_classes[] = 'with--' . $altered_region_id;
-      }
-      else {
-        $region_classes[$group_id][] = 'without--' . $altered_region_id;
-        $page_classes[] = 'without--' . $altered_region_id;
-      }
-    } 
-    // convert array of class names to string version
-    $pass_region_classes[$group_id]['#plain_text'] = implode(" ", $region_classes[$group_id]);
+  if ($layoutLayout) {
+    $region_groups = $layouts[$layoutLayout]['region_groups']['all'];
+    foreach($region_groups as $group_id => $group_info) {
+      $groupRegions = $region_groups[$group_id]['regions'];
+      // duplciate the #id into a class for better use with Omega.gs
+      $altered_group_id = str_replace("_", "-", $group_id);
+      $region_classes[$group_id][] = $altered_group_id . '-layout';
+      foreach($groupRegions as $region_id => $region_info) {
+        $altered_region_id = str_replace("_", "-", $region_id);
+        
+        if (isset($vars['page'][$region_id]['#region'])) {
+          $region_classes[$group_id][] = 'with--' . $altered_region_id;
+          $page_classes[] = 'with--' . $altered_region_id;
+        }
+        else {
+          $region_classes[$group_id][] = 'without--' . $altered_region_id;
+          $page_classes[] = 'without--' . $altered_region_id;
+        }
+      } 
+      // convert array of class names to string version
+      $pass_region_classes[$group_id]['#plain_text'] = implode(" ", $region_classes[$group_id]);
+    }
   }
   $pass_page_classes['#plain_text'] = implode(" ", $page_classes);
   // assign classes to page.tpl.php
