diff --git a/commerce_reports.info.yml b/commerce_reports.info.yml
index 62e0312..eca7568 100644
--- a/commerce_reports.info.yml
+++ b/commerce_reports.info.yml
@@ -1,8 +1,10 @@
 name: Commerce Reporting
 description: Provides reporting capabilities for sales, products, customers and payment gateways.
 package: Commerce (contrib)
+
 type: module
-core: 8.x
+core_version_requirement: ^8.8.0 || ^9.0
+
 dependencies:
   - commerce:commerce
   - commerce:commerce_price
diff --git a/composer.json b/composer.json
index ec4b60a..dcfb99a 100644
--- a/composer.json
+++ b/composer.json
@@ -5,6 +5,7 @@
   "homepage": "http://drupal.org/project/commerce_reports",
   "license": "GPL-2.0+",
   "require": {
-    "drupal/commerce": "^2.7"
+    "drupal/commerce": "^2.7",
+    "drupal/core": "^8.8 || ^9"
   }
 }
diff --git a/src/Plugin/views/field/ReportDateField.php b/src/Plugin/views/field/ReportDateField.php
index 008f03e..15355ba 100644
--- a/src/Plugin/views/field/ReportDateField.php
+++ b/src/Plugin/views/field/ReportDateField.php
@@ -4,7 +4,9 @@ namespace Drupal\commerce_reports\Plugin\views\field;

 use Drupal\Core\Datetime\DateFormatterInterface;
 use Drupal\Core\Datetime\Entity\DateFormat;
-use Drupal\Core\Entity\EntityManagerInterface;
+use Drupal\Core\Entity\EntityFieldManagerInterface;
+use Drupal\Core\Entity\EntityRepositoryInterface;
+use Drupal\Core\Entity\EntityTypeManagerInterface;
 use Drupal\Core\Field\FieldTypePluginManagerInterface;
 use Drupal\Core\Field\FormatterPluginManager;
 use Drupal\Core\Language\LanguageManagerInterface;
@@ -56,7 +58,7 @@ class ReportDateField extends EntityField {
    *   The plugin_id for the plugin instance.
    * @param mixed $plugin_definition
    *   The plugin implementation definition.
-   * @param \Drupal\Core\Entity\EntityManagerInterface $entity_manager
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entity_type_manager
    *   The field formatter plugin manager.
    * @param \Drupal\Core\Field\FormatterPluginManager $formatter_plugin_manager
    *   The field formatter plugin manager.
@@ -66,11 +68,15 @@ class ReportDateField extends EntityField {
    *   The language manager.
    * @param \Drupal\Core\Render\RendererInterface $renderer
    *   The renderer.
+   * @param \Drupal\Core\Entity\EntityRepositoryInterface $entity_repository
+   *   The entity repository.
+   * @param \Drupal\Core\Entity\EntityFieldManagerInterface $entity_field_manager
+   *   The entity field manager.
    * @param \Drupal\Core\Datetime\DateFormatterInterface $date_formatter
    *   The date formatter.
    */
-  public function __construct(array $configuration, $plugin_id, $plugin_definition, EntityManagerInterface $entity_manager, FormatterPluginManager $formatter_plugin_manager, FieldTypePluginManagerInterface $field_type_plugin_manager, LanguageManagerInterface $language_manager, RendererInterface $renderer, DateFormatterInterface $date_formatter) {
-    parent::__construct($configuration, $plugin_id, $plugin_definition, $entity_manager, $formatter_plugin_manager, $field_type_plugin_manager, $language_manager, $renderer);
+  public function __construct(array $configuration, $plugin_id, $plugin_definition, EntityTypeManagerInterface $entity_type_manager, FormatterPluginManager $formatter_plugin_manager, FieldTypePluginManagerInterface $field_type_plugin_manager, LanguageManagerInterface $language_manager, RendererInterface $renderer, EntityRepositoryInterface $entity_repository, EntityFieldManagerInterface $entity_field_manager, DateFormatterInterface $date_formatter) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition, $entity_type_manager, $formatter_plugin_manager, $field_type_plugin_manager, $language_manager, $renderer, $entity_repository, $entity_field_manager);
     $this->dateFormatter = $date_formatter;
   }

@@ -82,11 +88,13 @@ class ReportDateField extends EntityField {
       $configuration,
       $plugin_id,
       $plugin_definition,
-      $container->get('entity.manager'),
+      $container->get('entity_type.manager'),
       $container->get('plugin.manager.field.formatter'),
       $container->get('plugin.manager.field.field_type'),
       $container->get('language_manager'),
       $container->get('renderer'),
+      $container->get('entity.repository'),
+      $container->get('entity_field.manager'),
       $container->get('date.formatter')
     );
   }
@@ -100,7 +108,7 @@ class ReportDateField extends EntityField {
       }
       else {
         /* @var \Drupal\Core\Datetime\DateFormatInterface $formatter */
-        $formatter = $this->entityManager->getStorage('date_format')->load($this->dateFormat);
+        $formatter = $this->entityTypeManager->getStorage('date_format')->load($this->dateFormat);
         $this->dateFormatString = $formatter->getPattern();
       }
     }