diff --git a/src/Plugin/Commerce/ReportType/OrderReport.php b/src/Plugin/Commerce/ReportType/OrderReport.php
index 47c07be..2285fb3 100644
--- a/src/Plugin/Commerce/ReportType/OrderReport.php
+++ b/src/Plugin/Commerce/ReportType/OrderReport.php
@@ -57,12 +57,12 @@ class OrderReport extends ReportTypeBase {
   public function generateReports(OrderInterface $order) {
     $billing_profile = $order->getBillingProfile();
     /** @var \Drupal\address\Plugin\Field\FieldType\AddressItem $address */
-    $address = $billing_profile->get('address')->first();
+    $address = isset($billing_profile) ? $billing_profile->get('address')->first() : NULL;
     $values = [
       'order_type_id' => $order->bundle(),
       'amount' => $order->getTotalPrice(),
       'mail' => $order->getEmail(),
-      'billing_address' => $address->toArray(),
+      'billing_address' => isset($address) ? $address->toArray() : array(),
     ];
     $this->createFromOrder($order, $values);
   }
