diff --git a/modules/user/src/Form/UserPasswordResetForm.php b/modules/user/src/Form/UserPasswordResetForm.php
index a53c716e20..bcd0e185e6 100644
--- a/modules/user/src/Form/UserPasswordResetForm.php
+++ b/modules/user/src/Form/UserPasswordResetForm.php
@@ -47,7 +47,7 @@ class UserPasswordResetForm extends FormBase {
     }
     else {
       // No expiration for first time login.
-      $form['message'] = ['#markup' => $this->t('<p>This is a one-time login for %user_name.</p><p>Click on this button to log in to the site and change your password.</p>', ['%user_name' => $user->getAccountName()])];
+      $form['message'] = ['#markup' => $this->t('<p>This is a one-time login for %user_name.</p><p>Click on this button to log in to the site and set your new password.</p>', ['%user_name' => $user->getAccountName()])];
       $form['#title'] = $this->t('Set password');
     }