diff --git a/src/Form/MenuExportForm.php b/src/Form/MenuExportForm.php
index 2be8ef1..6984ebd 100644
--- a/src/Form/MenuExportForm.php
+++ b/src/Form/MenuExportForm.php
@@ -84,8 +84,13 @@ class MenuExportForm extends ConfigFormBase {
 			foreach($menuLinks as $link){
 				if(!empty($link)){
 					$linkArray = $link->toArray();
-					foreach($linkArray as $key=>$linkArrayItem){
-						$linkData[$key] = reset($linkArrayItem);
+					foreach($linkArray as $key=>$linkArrayItem) {
+						if ($key !== 'menu_item_roles') {
+						  $linkData[$key] = reset($linkArrayItem);
+            }
+            else {
+              $linkData[$key] = $linkArrayItem;
+            }
 					}
 					//$data[$link->id()] = $linkData;
 					$config->set($link->id(),$linkData);
