diff --git a/src/Form/BasicOverview.php b/src/Form/BasicOverview.php
index 2a5d4cb..9e5d7e8 100644
--- a/src/Form/BasicOverview.php
+++ b/src/Form/BasicOverview.php
@@ -144,6 +144,18 @@ class BasicOverview extends FormBase {
       ];
     }
 
+    if ($this->currentUser()->hasPermission('administer users')) {
+       $output['reset_skip_attempts'] = [
+          '#type' => 'submit',
+          '#value' => $this->t('Reset skip validation attempts'),
+          '#submit' => ['::resetSkipValidationAttempts'],
+       ];
+       $output['account'] = [
+           '#type' => 'value',
+           '#value' => $user,
+        ];
+    }
+
     return $output;
   }
 
@@ -184,4 +196,17 @@ class BasicOverview extends FormBase {
   public function submitForm(array &$form, FormStateInterface $form_state) {
   }
 
+  /**
+   *
+   */
+  public function resetSkipValidationAttempts($form, $form_state) {
+     $account = $form_state->getValue('account');
+     $tfa_data = $this->tfaGetTfaData($account->id(), $this->userData);
+     $tfa_data['validation_skipped'] = 0;
+     $this->tfaSaveTfaData($account->id(), $this->userData, $tfa_data);
+     drupal_set_message($this->t('Skipped validation attempts set to zero for user @name', [
+         '@name' => $account->getDisplayName(),
+     ]));
+  }
+
 }
diff --git a/src/Form/TfaLoginForm.php b/src/Form/TfaLoginForm.php
index f2654ef..dc2e8cc 100644
--- a/src/Form/TfaLoginForm.php
+++ b/src/Form/TfaLoginForm.php
@@ -216,24 +216,28 @@ class TfaLoginForm extends UserLoginForm {
    * @param \Drupal\Core\Form\FormStateInterface $form_state
    *   The state of the login form.
    */
-  public function loginWithoutTfa(FormStateInterface $form_state) {
-    // User may be able to skip TFA, depending on module settings and number of
-    // prior attempts.
-    $remaining = $this->tfaContext->remainingSkips();
-    if ($remaining) {
-      $user = $this->tfaContext->getUser();
-      $tfa_setup_link = Url::fromRoute('tfa.overview', [
-        'user' => $user->id(),
-      ])->toString();
-      drupal_set_message($this->t('You are required to setup two-factor authentication <a href="@link">here.</a> You have @remaining attempts left after this you will be unable to login.', [
-        '@remaining' => $remaining - 1,
-        '@link' => $tfa_setup_link,
-      ]), 'error');
-      $this->tfaContext->hasSkipped();
-      $this->tfaContext->doUserLogin();
-      $form_state->setRedirect('<front>');
+    public function loginWithoutTfa(FormStateInterface $form_state) {
+        // User may be able to skip TFA, depending on module settings and number of
+        // prior attempts.
+        $remaining = $this->tfaContext->remainingSkips();
+
+        if ($remaining) {
+            $user = $this->tfaContext->getUser();
+            $tfa_setup_link = Url::fromRoute('tfa.overview', [
+                'user' => $user->id(),
+            ])->toString();
+            $msg = $this->t('You are required to setup two-factor authentication. You have @remaining attempts left after this you will be unable to login.', [
+                '@remaining' => $remaining - 1,
+            ]);
+            \Drupal::messenger()->addMessage($msg,'error');
+
+            $this->tfaContext->hasSkipped();
+            $this->tfaContext->doUserLogin();
+            $form_state->setRedirect('<front>');
+        } else {
+            \Drupal::messenger()->addMessage($this->t('You cannot login because your account does not have a 2nd-Factor Authentication configured properly. Please contact the site administrator.'),'error', TRUE);
+        }
     }
-  }
 
   /**
    * Login submit handler for TFA form redirection.
